File: Algorithms/Dynamic Programming/memoization.py
def add_to_80(n):
    print('runs long time')
    return n + 80


def memoized_add_to_80():
    cache = {}
    def add_to_cache(n):
        if n not in cache:
            cache[n] = n + 80
            print('runs long time')
        return cache[n]
    return add_to_cache

print('First program')
print(add_to_80(5))
print(add_to_80(5))
print(add_to_80(5))


memoized = memoized_add_to_80()
print('Second program')
print(memoized(5))
print(memoized(5))
print(memoized(5))

File: Algorithms/Dynamic Programming/memoization.js
function addTo80(n) {
    console.log('Runs log time');
    return n + 80;
}


// using closure
function memoizedAddTo80() {
    let cache = {};
    return function(n) {
        if (!cache.n) {
            console.log('Runs log time');
            cache.n = n + 80;
        } return cache.n;
    }
}

console.log('First program');
console.log(addTo80(5));
console.log(addTo80(5));
console.log(addTo80(5));


const memoized = memoizedAddTo80();

console.log('Second program');
console.log(memoized(5));
console.log(memoized(5));
console.log(memoized(5));